File: volume_1/chapter_4/web.py
"""
web.py - Clean web interface for LLM testers
Now supports optional memory endpoints and session-based tracking.
"""

from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
import uvicorn
import io
from contextlib import redirect_stdout
import json

from stream import iter_text_chunks, normalize_response_text


class QueryRequest(BaseModel):
    topic: str
    provider: Optional[str] = None
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"


class QueryAllRequest(BaseModel):
    topic: str
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"


def _supports_memory(manager) -> bool:
    return bool(
        getattr(manager, "memory_enabled", False)
        and hasattr(manager, "get_history")
        and hasattr(manager, "reset_memory")
    )


def create_web_api(manager_class):
    app = FastAPI(
        title="LLM Service API",
        version="1.0.0",
        description="Universal API for LLM framework testing"
    )

    app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )

    manager = manager_class()

    @app.get("/")
    async def get_status():
        available = manager.get_available_providers()
        return {
            "framework": manager.framework,
            "available_providers": available,
            "total_available": len(available),
            "initialization_status": manager.initialization_messages,
            "status": "healthy" if available else "no_providers"
        }

    @app.get("/providers")
    async def get_providers():
        from utils import get_display_name, get_default_model
        providers = manager.get_available_providers()
        return {
            "framework": manager.framework,
            "providers": [
                {
                    "name": p,
                    "display_name": get_display_name(p),
                    "model": get_default_model(p),
                    "status": manager.initialization_messages.get(p, "Unknown")
                } for p in providers
            ],
            "count": len(providers)
        }

    @app.get("/capabilities")
    async def get_capabilities():
        return {
            "framework": manager.framework,
            "streaming": True,
            "stream_transport": "sse",
            "memory": _supports_memory(manager),
        }

    @app.post("/query")
    async def query_single(request: QueryRequest):
        try:
            with redirect_stdout(io.StringIO()):
                args = {
                    "topic": request.topic,
                    "provider": request.provider,
                    "template": request.template,
                    "max_tokens": request.max_tokens,
                    "temperature": request.temperature
                }
                if _supports_memory(manager):
                    args["session_id"] = request.session_id

                result = manager.ask_question(**args)

            if not result.get("success"):
                raise HTTPException(status_code=400, detail=result.get("error", "Query failed"))

            raw = result.get("response")
            content = raw.content if hasattr(raw, "content") else raw

            return {
                "success": True,
                "framework": manager.framework,
                "provider": result["provider"],
                "model": result["model"],
                "response": content,
                "parameters": {
                    "temperature": result["temperature"],
                    "max_tokens": result["max_tokens"],
                    "template": request.template
                },
                "prompt": result["prompt"],
                "session_id": result.get("session_id", "default")
            }

        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @app.post("/query-stream")
    async def query_stream(request: QueryRequest):
        async def stream_events():
            try:
                with redirect_stdout(io.StringIO()):
                    args = {
                        "topic": request.topic,
                        "provider": request.provider,
                        "template": request.template,
                        "max_tokens": request.max_tokens,
                        "temperature": request.temperature,
                    }
                    if _supports_memory(manager):
                        args["session_id"] = request.session_id

                    result = manager.ask_question(**args)

                if not result.get("success"):
                    error_payload = {"type": "error", "error": result.get("error", "Query failed")}
                    yield f"data: {json.dumps(error_payload)}\n\n"
                    return

                response_text = normalize_response_text(result.get("response"))
                async for chunk in iter_text_chunks(response_text):
                    payload = {"type": "chunk", "content": chunk}
                    yield f"data: {json.dumps(payload)}\n\n"

                done_payload = {
                    "type": "done",
                    "provider": result.get("provider"),
                    "model": result.get("model"),
                }
                yield f"data: {json.dumps(done_payload)}\n\n"
            except Exception as e:
                payload = {"type": "error", "error": str(e)}
                yield f"data: {json.dumps(payload)}\n\n"

        return StreamingResponse(stream_events(), media_type="text/event-stream")

    @app.post("/query-all")
    async def query_all(request: QueryAllRequest):
        try:
            with redirect_stdout(io.StringIO()):
                result = manager.query_all_providers(
                    topic=request.topic,
                    template=request.template,
                    max_tokens=request.max_tokens,
                    temperature=request.temperature
                )

            if not result.get("success"):
                raise HTTPException(status_code=400, detail=result.get("error", "Query failed"))

            clean_responses = {}
            for provider, res in result["responses"].items():
                raw = res.get("response")
                content = raw.content if hasattr(raw, "content") else raw
                clean_responses[provider] = {
                    "success": res["success"],
                    "model": res.get("model", ""),
                    "response": content,
                    "parameters": {
                        "temperature": res.get("temperature"),
                        "max_tokens": res.get("max_tokens")
                    }
                }

            return {
                "success": True,
                "framework": manager.framework,
                "prompt": result["prompt"],
                "responses": clean_responses
            }

        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @app.get("/history")
    async def get_history(provider: str, session_id: str = "default"):
        if not _supports_memory(manager):
            raise HTTPException(status_code=400, detail="Memory not supported by this manager")
        return manager.get_history(provider, session_id)

    @app.post("/reset-memory")
    async def reset_memory(provider: Optional[str] = None, session_id: Optional[str] = None):
        if not _supports_memory(manager):
            raise HTTPException(status_code=400, detail="Memory not supported by this manager")
        return manager.reset_memory(provider, session_id)

    return app


def run_web_server(manager_class, host: str = "0.0.0.0", port: int = 8000):
    app = create_web_api(manager_class)
    try:
        framework_name = manager_class().framework
    except Exception:
        framework_name = "Unknown"

    print(f"Starting web server for {framework_name}")
    print(f"Docs: http://{host}:{port}/docs")
    print(f"Health: http://{host}:{port}/")
    uvicorn.run(app, host=host, port=port)


def main():
    print("Universal LLM Web API")
    print("Run using `run_web_server(manager_class)`")


if __name__ == "__main__":
    main()

File: volume_1/chapter_4/web.js
/**
 * web.js - Clean web interface for JavaScript LLM testers.
 * Supports optional memory endpoints and session-based tracking.
 */

import express from 'express';
import cors from 'cors';
import { getDisplayName, getDefaultModel } from './utils.js';
import { chunkText, normalizeResponseText } from './stream.js';

async function captureConsoleOutputAsync(fn) {
    const originalLog = console.log;
    const logs = [];
    console.log = (...args) => logs.push(args.join(' '));
    try {
        const result = await fn();
        return { result, logs: logs.join('\n') };
    } finally {
        console.log = originalLog;
    }
}

function supportsMemory(manager) {
    return Boolean(
        manager?.memoryEnabled
        && typeof manager.getHistory === 'function'
        && typeof manager.resetMemory === 'function'
    );
}

function buildManager(managerClassOrFactory) {
    if (typeof managerClassOrFactory !== 'function') {
        throw new Error('Invalid manager class/factory provided');
    }
    try {
        return new managerClassOrFactory();
    } catch {
        return managerClassOrFactory();
    }
}

function createWebApi(managerClassOrFactory) {
    const app = express();
    app.use(cors());
    app.use(express.json());

    let manager;
    const initPromise = (async () => {
        manager = buildManager(managerClassOrFactory);
        await manager._checkProviders();
        return manager;
    })();

    app.use(async (_, __, next) => {
        if (!manager) {
            await initPromise;
        }
        next();
    });

    app.get('/', async (_, res) => {
        const available = manager.getAvailableProviders();
        res.json({
            framework: manager.framework,
            available_providers: available,
            total_available: available.length,
            initialization_status: manager.initializationMessages,
            status: available.length > 0 ? 'healthy' : 'no_providers',
        });
    });

    app.get('/providers', async (_, res) => {
        const available = manager.getAvailableProviders();
        res.json({
            framework: manager.framework,
            providers: available.map((provider) => ({
                name: provider,
                display_name: getDisplayName(provider),
                model: getDefaultModel(provider),
                status: manager.initializationMessages[provider] || 'Unknown',
            })),
            count: available.length,
        });
    });

    app.get('/capabilities', async (_, res) => {
        res.json({
            framework: manager.framework,
            streaming: true,
            stream_transport: 'sse',
            memory: supportsMemory(manager),
        });
    });

    app.post('/query', async (req, res) => {
        try {
            const {
                topic,
                provider = null,
                template = '{topic}',
                max_tokens = 1000,
                temperature = 0.7,
                session_id = 'default',
            } = req.body;

            if (!topic) {
                return res.status(400).json({ error: 'Topic is required' });
            }

            const { result, logs } = await captureConsoleOutputAsync(async () => {
                if (supportsMemory(manager)) {
                    return manager.askQuestion(topic, provider, template, max_tokens, temperature, session_id);
                }
                return manager.askQuestion(topic, provider, template, max_tokens, temperature);
            });

            if (!result.success) {
                return res.status(400).json({
                    error: result.error || 'Query failed',
                    provider: result.provider,
                    debug: logs || null,
                });
            }

            return res.json({
                success: true,
                framework: manager.framework,
                provider: result.provider,
                model: result.model,
                response: result.response,
                parameters: {
                    temperature: result.temperature,
                    max_tokens: result.maxTokens,
                    template,
                },
                prompt: result.prompt,
                session_id: result.sessionId || 'default',
                ...(logs ? { debug: logs } : {}),
            });
        } catch (error) {
            return res.status(500).json({ error: error.message, framework: manager.framework });
        }
    });

    app.post('/query-stream', async (req, res) => {
        try {
            const {
                topic,
                provider = null,
                template = '{topic}',
                max_tokens = 1000,
                temperature = 0.7,
                session_id = 'default',
            } = req.body;

            if (!topic) {
                return res.status(400).json({ error: 'Topic is required' });
            }

            const { result } = await captureConsoleOutputAsync(async () => {
                if (supportsMemory(manager)) {
                    return manager.askQuestion(topic, provider, template, max_tokens, temperature, session_id);
                }
                return manager.askQuestion(topic, provider, template, max_tokens, temperature);
            });

            res.setHeader('Content-Type', 'text/event-stream');
            res.setHeader('Cache-Control', 'no-cache');
            res.setHeader('Connection', 'keep-alive');

            if (!result.success) {
                res.write(`data: ${JSON.stringify({ type: 'error', error: result.error || 'Query failed' })}\n\n`);
                return res.end();
            }

            const responseText = normalizeResponseText(result.response);
            for (const chunk of chunkText(responseText)) {
                res.write(`data: ${JSON.stringify({ type: 'chunk', content: chunk })}\n\n`);
            }

            res.write(`data: ${JSON.stringify({ type: 'done', provider: result.provider, model: result.model })}\n\n`);
            return res.end();
        } catch (error) {
            if (!res.headersSent) {
                return res.status(500).json({ error: error.message, framework: manager.framework });
            }
            res.write(`data: ${JSON.stringify({ type: 'error', error: error.message })}\n\n`);
            return res.end();
        }
    });

    app.post('/query-all', async (req, res) => {
        try {
            const { topic, template = '{topic}', max_tokens = 1000, temperature = 0.7 } = req.body;
            if (!topic) {
                return res.status(400).json({ error: 'Topic is required' });
            }

            const { result, logs } = await captureConsoleOutputAsync(async () => (
                manager.queryAllProviders(topic, template, max_tokens, temperature)
            ));

            if (!result.success) {
                return res.status(400).json({
                    error: result.error || 'Query failed',
                    framework: manager.framework,
                    debug: logs || null,
                });
            }

            const cleanResponses = {};
            let successful = 0;
            let failed = 0;
            for (const [provider, response] of Object.entries(result.responses || {})) {
                if (response.success) {
                    cleanResponses[provider] = {
                        success: true,
                        response: response.response,
                        model: response.model,
                        parameters: {
                            temperature: response.temperature,
                            max_tokens: response.maxTokens,
                        },
                    };
                    successful += 1;
                } else {
                    cleanResponses[provider] = {
                        success: false,
                        error: response.error || 'Unknown error',
                        model: response.model || 'unknown',
                    };
                    failed += 1;
                }
            }

            return res.json({
                success: true,
                framework: manager.framework,
                prompt: result.prompt,
                responses: cleanResponses,
                summary: {
                    total_providers: Object.keys(result.responses || {}).length,
                    successful,
                    failed,
                },
                parameters: { temperature, max_tokens, template },
                ...(logs ? { debug: logs } : {}),
            });
        } catch (error) {
            return res.status(500).json({ error: error.message, framework: manager.framework });
        }
    });

    app.get('/history', async (req, res) => {
        if (!supportsMemory(manager)) {
            return res.status(400).json({ error: 'Memory not supported by this manager' });
        }
        const { provider, session_id = 'default' } = req.query;
        if (!provider) {
            return res.status(400).json({ error: 'provider is required' });
        }
        return res.json(manager.getHistory(provider, session_id));
    });

    app.post('/reset-memory', async (req, res) => {
        if (!supportsMemory(manager)) {
            return res.status(400).json({ error: 'Memory not supported by this manager' });
        }
        const body = req.body || {};
        const bodyProvider = body.provider ?? null;
        const bodySessionId = body.session_id ?? null;
        const queryProvider = req.query?.provider ?? null;
        const querySessionId = req.query?.session_id ?? null;
        const provider = bodyProvider !== null ? bodyProvider : queryProvider;
        const sessionId = bodySessionId !== null ? bodySessionId : querySessionId;
        return res.json(manager.resetMemory(provider, sessionId));
    });

    app.get('/health', async (_, res) => {
        const available = manager.getAvailableProviders();
        res.json({
            status: available.length > 0 ? 'healthy' : 'unhealthy',
            framework: manager.framework,
            providers_available: available.length,
        });
    });

    return app;
}

export async function runWebServer(managerClassOrFactory, host = '0.0.0.0', port = 8000) {
    const app = createWebApi(managerClassOrFactory);

    let frameworkName = 'Unknown';
    try {
        frameworkName = buildManager(managerClassOrFactory).framework;
    } catch {
        // Ignore display failure
    }

    app.listen(port, host, () => {
        console.log(`Starting web server for ${frameworkName} framework...`);
        console.log(`Health check: http://${host}:${port}/health`);
        console.log(`Status: http://${host}:${port}/`);
    });
}

function main() {
    console.log('Universal LLM Web API (JavaScript)');
}

if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}
